bson/bsonrw/value_reader.go

```diff
@@ -748,7 +748,11 @@ func (vr *valueReader) readBytes(length int32) ([]byte, error) {
 
 	start := vr.offset
 	vr.offset += int64(length)
-	return vr.d[start : start+int64(length)], nil
+
+	b := make([]byte, length)
+	copy(b, vr.d[start:start+int64(length)])
```

Comment:
Will this copying significantly increase allocations during `Unmarshal`ing?

Reply:
It is possible, although it seems like it should not be noticeable in most cases. A case where it could be noticeable is if a user is storing large binary blobs in BSON documents.

In any case, this is necessary to maintain the behavior expectation of `Unmarshal` (i.e. modifying the input after unmarshaling has no affect on the unmarshaled values).

Reply:
Makes sense; thanks.

Reply:
Not sure if this is what you meant, but I realized copying the byte slice in `vr.readBytes()` is unnecessary and we can copy the value in `vr.ReadBinary()` instead. All other functions that call `vr.readBytes()` already copy or convert the returned byte slice. I made a PR for that improvement here: https://github.com/mongodb/mongo-go-driver/pull/891